File: 338. Counting Bits/Solution.py
class Solution:
    def countBits(self, n: int) -> List[int]:
        if n == 0: return [0]
        if n == 1: return [0, 1]
        res = [0 for i in range(n + 1)]
        res[1] = 1
        left, gap = 1, 1
        for right in range(2, n + 1):
            if right == left + gap:
                res[right] = 1
                left = right
                gap *= 2
            else:
                res[right] = res[right - gap] + 1
        return res

File: 338. Counting Bits/Solution.cpp
class Solution {
public:
    vector<int> countBits(int n) {
        if (n == 0) return vector<int> { 0 };
        if (n == 1) return vector<int> { 0, 1 };
        vector<int> res(n + 1, 0);
        res[1] = 1;
        int left = 1, gap = 1;
        for (int right = 2; right <= n; right++) {
            if (right == left + gap) {
                res[right] = 1;
                left = right;
                gap *= 2;
            } else {
                res[right] = res[right - gap] + 1;
            }
        }
        return res;
    }
};
